# Data Preparation

<hr style="border:2px solid red"> </hr>

Now that I've got the data, I have to clean it in a way that I can explore and model on it. This will require an understanding of what the columns represent, versus what I will need in the final product.

In [1]:
# Data Science Libraries
import pandas as pd
import numpy as np
import datetime

# Import my own functions
import wrangle

# Block Warning Boxes
import warnings
warnings.filterwarnings("ignore")

# Remove Limits On Viewing Dataframes
# pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', None)

In response to my time constraint for this project, I will reduce my data to just twins_raw data for now. I will keep the code for exploring all the data, including `twins_rawevent` and `twins_config` for another time.

#### TWINS raw data
>This data product contains TWINS raw science data downloaded from the spacecraft in
continuous mode. This includes Air Temperature Sensor PT1000 raw values, Wind Sensor
counters and temperatures, and ASIC temperature. 

In [2]:
# Acquiring the data using functions in wrangle
df = wrangle.twins_raw_data()

Reading Data from local file...


In [3]:
# How big is this dataframe?
df.shape

(2350436, 63)

In [4]:
# What does it look like?
df.head()

AOBT          SCLK                LMST            LTST  \
0  596876952.0  5.968612e+08  00004M06:46:33.826  00004 06:05:41   
1  596876953.0  5.968612e+08  00004M06:46:34.799  00004 06:05:42   
2  596876954.0  5.968612e+08  00004M06:46:35.772  00004 06:05:43   
3  596876955.0  5.968612e+08  00004M06:46:36.746  00004 06:05:44   
4  596876956.0  5.968612e+08  00004M06:46:37.719  00004 06:05:45   

                      UTC  BMY_2L_TEMP_1  BMY_2L_TEMP_2  BMY_2L_TEMP_3  \
0  2018-334T14:46:55.755Z        -4353.0        -4645.0        -4778.0   
1  2018-334T14:46:56.755Z        -4415.0        -4723.0        -4870.0   
2  2018-334T14:46:57.755Z        -4424.0        -4727.0        -4861.0   
3  2018-334T14:46:58.755Z        -4409.0        -4722.0        -4859.0   
4  2018-334T14:46:59.755Z        -4424.0        -4728.0        -4873.0   

   BMY_2L_TEMP_4  BMY_2L_TEMP_4_AVERAGE  ...  BPY_WD_OUT_6  BPY_WD_OUT_7  \
0        -5006.0                    NaN  ...           NaN           NaN   
1        -5099.0                    NaN  ...           NaN           NaN   
2        -5107.0                    NaN  ...           NaN           NaN   
3        -5109.0                    NaN  ...           NaN           NaN   
4        -5107.0                    NaN  ...           NaN           NaN   

   BPY_WD_OUT_8  BPY_WD_OUT_9  BPY_WD_OUT_10  BPY_WD_OUT_11  BPY_WD_OUT_12  \
0           NaN           NaN            NaN            NaN            NaN   
1           NaN           NaN            NaN            NaN            NaN   
2           NaN           NaN            NaN            NaN            NaN   
3           NaN           NaN            NaN            NaN            NaN   
4           NaN           NaN            NaN            NaN            NaN   

   BPY_WIND_FREQUENCY  BPY_AIR_TEMP_FREQUENCY  BPY_ASIC_TEMP  
0                 NaN                     NaN            NaN  
1                 NaN                     NaN            NaN  
2                 NaN                     NaN            NaN  
3                 NaN                     NaN            NaN  
4                 NaN                     NaN            NaN  

[5 rows x 63 columns]

In [5]:
# What do my columns look like
df.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2350436 entries, 0 to 4443
Data columns (total 63 columns):
 #   Column                  Non-Null Count    Dtype  
---  ------                  --------------    -----  
 0   AOBT                    2350436 non-null  float64
 1   SCLK                    2350436 non-null  float64
 2   LMST                    2350436 non-null  object 
 3   LTST                    2350436 non-null  object 
 4   UTC                     2350436 non-null  object 
 5   BMY_2L_TEMP_1           1440001 non-null  float64
 6   BMY_2L_TEMP_2           1440001 non-null  float64
 7   BMY_2L_TEMP_3           1440001 non-null  float64
 8   BMY_2L_TEMP_4           707662 non-null   float64
 9   BMY_2L_TEMP_4_AVERAGE   732324 non-null   float64
 10  BMY_2L_TEMP_4_STD       732324 non-null   float64
 11  BMY_2L_TEMP_5           1440001 non-null  float64
 12  BMY_2L_TEMP_6           1440001 non-null  float64
 13  BMY_AIR_TEMP            707662 non-null   float64
 14  BMY_A

In [6]:
# I can see I have some missing values, let's see how many in each column:
df.isnull().sum()

AOBT                           0
SCLK                           0
LMST                           0
LTST                           0
UTC                            0
                           ...  
BPY_WD_OUT_11             911195
BPY_WD_OUT_12             911195
BPY_WIND_FREQUENCY        911195
BPY_AIR_TEMP_FREQUENCY    911195
BPY_ASIC_TEMP             911210
Length: 63, dtype: int64

Looks like I have a lot of missing data to look into. Before I do that tho I think I'll drop some columns first.  
Now I'd like to drop some columns so my data is simple to explore, however I think I should create a data dictionary first so I don't delete anything important!

## TWINS Raw Data Columns
|System          |  # | Column  |Data Type              | Description                                         |
|:---------------|:---|:--------|:----------------------|:----------------------------------------------------|
|Time References |  1 | AOBT    |ASCII_Real             | APSS Onboard Time
|                |  2 | SCLK    |ASCII_Real             | Spacecraft Clock
|                |  3 | LMST    |ASCII_String           | Local Mean Solar Time
|                |  4 | LTST    |ASCII_String           | Local True Solar Time
|                |  5 | UTC     |ASCII_Date_Time_DOY_UTC| Coordinated Universal Time
|================|====|=========|=======================|=====================================================|
| BOOM -Y        |  6 | BMY_2L_TEMP_1 | ASCII_Integer | Wind Sensor transducer 1 Printed Circuit Board temperature PT-1000 Platinum Resistance Thermometer |
|                |  7 | BMY_2L_TEMP_2 | ASCII_Integer | WS transducer 2 PCB temperature PT-1000 PRT | 
|                |  8 | BMY_2L_TEMP_3 | ASCII_Integer | WS transducer 3 PCB temperature PT-1000 PRT |
|                |  9 | BMY_2L_TEMP_4 | ASCII_Integer | ATS-mid-rodtemperature: PT1000 PRT sensor located at an intermediate position in the ATS rod|
|                | 10 | BMY_2L_TEMP_4_AVERAGE | ASCII_Integer | ATS-mid-rod temperature average of the last N samples
|                | 11 | BMY_2L_TEMP_4_STD     | ASCII_Integer | ATS-mid-rod temperature standard deviation of the last N samples
|                | 12 | BMY_2L_TEMP_5         | ASCII_Integer | Boom Housing Temp: PT-1000 PRT located at the Boom housing near the base of the ATS rod
|                | 13 | BMY_2L_TEMP_6         | ASCII_Integer | Calibration resistor: 1K ohm
|                | 14 | BMY_AIR_TEMP | ASCII_Integer | ATS-rod-extreme temperature: PT1000 PRT located at ATS extreme
|                | 15 | BMY_AIR_TEMP_AVERAGE  | ASCII_Integer | ATS-rod-extreme temperature average of the last N samples
|                | 16 | BMY_AIR_TEMP_STD | ASCII_Integer | ATS-rod-extreme temperature standard deviation of the last N samples
|                | 17 | BMY_WD_REF_OUT_1 | ASCII_Integer | WS transducer 1 cold die temperature
|                | 18 | BMY_WD_REF_OUT_2 | ASCII_Integer | WS transducer 2 cold die temperature
|                | 19 | BMY_WD_REF_OUT_3 | ASCII_Integer | WS transducer 3 cold die temperature
|                | 20 | BMY_WD_OUT_1     | ASCII_Integer | Number of counts measured for WS channel 1
|                | 21 | BMY_WD_OUT_2     | ASCII_Integer | Number of counts measured for WS channel 2
|                | 22 | BMY_WD_OUT_3     | ASCII_Integer | Number of counts measured for WS channel 3
|                | 23 | BMY_WD_OUT_4     | ASCII_Integer | Number of counts measured for WS channel 4
|                | 24 | BMY_WD_OUT_5     | ASCII_Integer | Number of counts measured for WS channel 5
|                | 25 | BMY_WD_OUT_6     | ASCII_Integer | Number of counts measured for WS channel 6
|                | 26 | BMY_WD_OUT_7     | ASCII_Integer | Number of counts measured for WS channel 7
|                | 27 | BMY_WD_OUT_8     | ASCII_Integer | Number of counts measured for WS channel 8
|                | 28 | BMY_WD_OUT_9     | ASCII_Integer | Number of counts measured for WS channel 9
|                | 29 | BMY_WD_OUT_10    | ASCII_Integer | Number of counts measured for WS channel 10
|                | 30 | BMY_WD_OUT_11    | ASCII_Integer | Number of counts measured for WS channel 11
|                | 31 | BMY_WD_OUT_12 | ASCII_Integer | Number of counts measured for WS channel 12
|                | 32 | BMY_ASIC_TEMP | ASCII_Integer | ASIC temperature
|                | 33 | BMY_AIR_TEMP_FREQUENCY | ASCII_String | Air temperature channels frequency or frequencies
|                | 34 | BMY_WIND_FREQUENCY | ASCII_String | Wind channels frequency or frequencies
|================|====|=========|=======================|=====================================================|
| BOOM +Y        | 35 | BPY_2L_TEMP_1 | ASCII_Integer | WS transducer 1 PCB temperature PT-1000 PRT
|                | 36 | BPY_2L_TEMP_2 | ASCII_Integer | WS transducer 2 PCB temperature PT-1000 PRT
|                | 37 | BPY_2L_TEMP_3 | ASCII_Integer | WS transducer 3 PCB temperature PT-1000 PRT
|                | 38 | BPY_2L_TEMP_4 | ASCII_Integer | Calibration resistor: 1K ohm
|                | 39 | BPY_2L_TEMP_5 | ASCII_Integer | ATS-mid-rod temperature: PT1000 PRT sensor located at a intermediate position in the ATS rod
|                | 40 | BPY_2L_TEMP_5_AVERAGE | ASCII_Integer | ATS-mid-rod temperature average of the last N samples
|                | 41 | BPY_2L_TEMP_5_STD | ASCII_Integer | ATS-mid-rod temperature standard deviation of the last N samples
|                | 42 | BPY_2L_TEMP_6 | ASCII_Integer | Boom Housing Temp: PT-1000 PRT located at the Boom housing near the base of the ATS rod
|                | 43 | BPY_AIR_TEMP | ASCII_Integer | ATS-rod-extreme temperature: PT1000 PRT located at ATS extreme
|                | 44 | BPY_AIR_TEMP_AVERAGE | ASCII_Integer | ATS-rod-extreme temperature average of the last N samples
|                | 45 | BPY_AIR_TEMP_STD | ASCII_Integer | ATS-rod-extreme temperature standard deviation of the last N samples
|                | 46 | BPY_WD_REF_OUT_1 | ASCII_Integer | WS transducer 1 cold die temperature
|                | 47 | BPY_WD_REF_OUT_2 | ASCII_Integer | WS transducer 2 cold die temperature
|                | 48 | BPY_WD_REF_OUT_3 | ASCII_Integer | WS transducer 3 cold die temperature
|                | 49 | BPY_WD_OUT_1     | ASCII_Integer | Number of counts measured for WS channel 1
|                | 50 | BPY_WD_OUT_2     | ASCII_Integer | Number of counts measured for WS channel 2
|                | 51 | BPY_WD_OUT_3     | ASCII_Integer | Number of counts measured for WS channel 3
|                | 52 | BPY_WD_OUT_4     | ASCII_Integer | Number of counts measured for WS channel 4
|                | 53 | BPY_WD_OUT_5     | ASCII_Integer | Number of counts measured for WS channel 5
|                | 54 | BPY_WD_OUT_6     | ASCII_Integer | Number of counts measured for WS channel 6
|                | 55 | BPY_WD_OUT_7     | ASCII_Integer | Number of counts measured for WS channel 7
|                | 56 | BPY_WD_OUT_8     | ASCII_Integer | Number of counts measured for WS channel 8
|                | 57 | BPY_WD_OUT_9     | ASCII_Integer | Number of counts measured for WS channel 9
|                | 58 | BPY_WD_OUT_10    | ASCII_Integer | Number of counts measured for WS channel 10
|                | 59 | BPY_WD_OUT_11    | ASCII_Integer | Number of counts measured for WS channel 11
|                | 60 | BPY_WD_OUT_12    | ASCII_Integer | Number of counts measured for WS channel 12
|                | 61 | BPY_ASIC_TEMP    | ASCII_Integer | ASIC temperature
|                | 62 | BPY_AIR_TEMP_FREQUENCY | ASCII_String |Air temperature channels frequency or frequencies
|                | 63 | BPY_WIND_FREQUENCY | ASCII_String | Wind channels frequency or frequencies

In [7]:
# df.BPY_AIR_TEMP_FREQUENCY

#### TWINS RAW Data
>In continuous mode, and for mid-rod and extreme-rod air temperature PT1000 sensors, the
recorded value is an average and a standard deviation using a window of N samples (N
defined to 7). This window starts at the time in which the decimation process selects the
rest of the data, each M samples. This means that the actual averages and standard
deviations are obtained N seconds after the rest of the samples. For convenience, the data
products contain all values aligned at the same time tags.  
Air Temperature data for which the average and standard deviations are calculated have
three columns: one for the raw value, for modes in which a single sample was retrieved,
and two others with the average and standard deviations. Depending on the downlink mode
used to retrieve the data, the appropriate column will contain the value and the other(s) will
be empty.

## My plan
BOOM -Y , Take average of all these temps and make one column
- BMY_2L_TEMP_1
- BMY_2L_TEMP_2
- BMY_2L_TEMP_3
- BMY_2L_TEMP_4
- BMY_2L_TEMP_4_AVERAGE [DROP]
- BPY_2L_TEMP_5
- BPY_2L_TEMP_5_AVERAGE [DROP]
- BPY_2L_TEMP_5_STD [DROP]
- BPY_2L_TEMP_6
---
- BMY_AIR_TEMP
- BMY_AIR_TEMP_AVERAGE [DROP]
- BMY_AIR_TEMP_STD [DROP]
---
Take the average of these
- BMY_WD_REF_OUT_1
- BMY_WD_REF_OUT_2
- BMY_WD_REF_OUT_3
---
Take the average of these
- BPY_WD_OUT_1
- BPY_WD_OUT_2
- BPY_WD_OUT_3
- BPY_WD_OUT_4
- BPY_WD_OUT_5
- BPY_WD_OUT_6
- BPY_WD_OUT_7
- BPY_WD_OUT_8
- BPY_WD_OUT_9
- BPY_WD_OUT_10
- BPY_WD_OUT_11
- BPY_WD_OUT_12
---
- BMY_ASIC_TEMP
- BPY_AIR_TEMP_FREQUENCY
- BPY_WIND_FREQUENCY

BOOM +Y
- BPY_2L_TEMP_1
- BPY_2L_TEMP_2
- BPY_2L_TEMP_3
- BPY_2L_TEMP_4
- BPY_2L_TEMP_5
- BPY_2L_TEMP_5_AVERAGE
- BPY_2L_TEMP_5_STD
- BPY_2L_TEMP_6
---
- BPY_AIR_TEMP
- BPY_AIR_TEMP_AVERAGE
- BPY_AIR_TEMP_STD
---
- BPY_WD_REF_OUT_1
- BPY_WD_REF_OUT_2
- BPY_WD_REF_OUT_3
---
- BPY_WD_OUT_1
- BPY_WD_OUT_2
- BPY_WD_OUT_3
- BPY_WD_OUT_4
- BPY_WD_OUT_5
- BPY_WD_OUT_6
- BPY_WD_OUT_7
- BPY_WD_OUT_8
- BPY_WD_OUT_9
- BPY_WD_OUT_10
- BPY_WD_OUT_11
- BPY_WD_OUT_12
---
- BPY_ASIC_TEMP
- BPY_AIR_TEMP_FREQUENCY
- BPY_WIND_FREQUENCY

In [7]:
df.head()

AOBT          SCLK                LMST            LTST  \
0  596876952.0  5.968612e+08  00004M06:46:33.826  00004 06:05:41   
1  596876953.0  5.968612e+08  00004M06:46:34.799  00004 06:05:42   
2  596876954.0  5.968612e+08  00004M06:46:35.772  00004 06:05:43   
3  596876955.0  5.968612e+08  00004M06:46:36.746  00004 06:05:44   
4  596876956.0  5.968612e+08  00004M06:46:37.719  00004 06:05:45   

                      UTC  BMY_2L_TEMP_1  BMY_2L_TEMP_2  BMY_2L_TEMP_3  \
0  2018-334T14:46:55.755Z        -4353.0        -4645.0        -4778.0   
1  2018-334T14:46:56.755Z        -4415.0        -4723.0        -4870.0   
2  2018-334T14:46:57.755Z        -4424.0        -4727.0        -4861.0   
3  2018-334T14:46:58.755Z        -4409.0        -4722.0        -4859.0   
4  2018-334T14:46:59.755Z        -4424.0        -4728.0        -4873.0   

   BMY_2L_TEMP_4  BMY_2L_TEMP_4_AVERAGE  ...  BPY_WD_OUT_6  BPY_WD_OUT_7  \
0        -5006.0                    NaN  ...           NaN           NaN   
1        -5099.0                    NaN  ...           NaN           NaN   
2        -5107.0                    NaN  ...           NaN           NaN   
3        -5109.0                    NaN  ...           NaN           NaN   
4        -5107.0                    NaN  ...           NaN           NaN   

   BPY_WD_OUT_8  BPY_WD_OUT_9  BPY_WD_OUT_10  BPY_WD_OUT_11  BPY_WD_OUT_12  \
0           NaN           NaN            NaN            NaN            NaN   
1           NaN           NaN            NaN            NaN            NaN   
2           NaN           NaN            NaN            NaN            NaN   
3           NaN           NaN            NaN            NaN            NaN   
4           NaN           NaN            NaN            NaN            NaN   

   BPY_WIND_FREQUENCY  BPY_AIR_TEMP_FREQUENCY  BPY_ASIC_TEMP  
0                 NaN                     NaN            NaN  
1                 NaN                     NaN            NaN  
2                 NaN                     NaN            NaN  
3                 NaN                     NaN            NaN  
4                 NaN                     NaN            NaN  

[5 rows x 63 columns]

During science monitoring phase, after all instruments (SEIS, HP3) are deployed, only one
boom will be used at a time. A selection of which boom is operative (measuring) will be
done in accordance to where the wind flows come from, highly determined by the time of
the Martian day (among other factors). The boom selection pattern will be commanded
once per week, based on past observations and predictive models

In [8]:
# If I'm being honest there's a lot of overwhelming data here
# So I'm gonna break it down...
# I'll start by making time into a dataframe
time = df[['AOBT', 'SCLK', 'LMST', 'LTST', 'UTC']].head()
time.head()

AOBT          SCLK                LMST            LTST  \
0  596876952.0  5.968612e+08  00004M06:46:33.826  00004 06:05:41   
1  596876953.0  5.968612e+08  00004M06:46:34.799  00004 06:05:42   
2  596876954.0  5.968612e+08  00004M06:46:35.772  00004 06:05:43   
3  596876955.0  5.968612e+08  00004M06:46:36.746  00004 06:05:44   
4  596876956.0  5.968612e+08  00004M06:46:37.719  00004 06:05:45   

                      UTC  
0  2018-334T14:46:55.755Z  
1  2018-334T14:46:56.755Z  
2  2018-334T14:46:57.755Z  
3  2018-334T14:46:58.755Z  
4  2018-334T14:46:59.755Z

In [9]:
# Now to separate and examine the BMY_2L_TEMP data
# Why does TEMP 6 have positive numbers?
BMY_2L_Temp = df[['BMY_2L_TEMP_1', 'BMY_2L_TEMP_2', 'BMY_2L_TEMP_3', 'BMY_2L_TEMP_4', 'BMY_2L_TEMP_5', 'BMY_2L_TEMP_6']]
BMY_2L_Temp.head()

BMY_2L_TEMP_1  BMY_2L_TEMP_2  BMY_2L_TEMP_3  BMY_2L_TEMP_4  BMY_2L_TEMP_5  \
0        -4353.0        -4645.0        -4778.0        -5006.0        -4512.0   
1        -4415.0        -4723.0        -4870.0        -5099.0        -4600.0   
2        -4424.0        -4727.0        -4861.0        -5107.0        -4613.0   
3        -4409.0        -4722.0        -4859.0        -5109.0        -4603.0   
4        -4424.0        -4728.0        -4873.0        -5107.0        -4610.0   

   BMY_2L_TEMP_6  
0         1002.0  
1          936.0  
2          930.0  
3          916.0  
4          922.0

In [10]:
# I know it's simple but I want to separate EVERYTHING
BMY_AIR_TEMP = df.BMY_AIR_TEMP
BMY_AIR_TEMP.head()

0   -5703.0
1   -5776.0
2   -5803.0
3   -5771.0
4   -5754.0
Name: BMY_AIR_TEMP, dtype: float64

In [11]:
# Boom Minus Y reference out = Wind Speed transducer cold die temperature
BMY_WD_REF_OUT = df[['BMY_WD_REF_OUT_1', 'BMY_WD_REF_OUT_2', 'BMY_WD_REF_OUT_3']]
BMY_WD_REF_OUT.head()

BMY_WD_REF_OUT_1  BMY_WD_REF_OUT_2  BMY_WD_REF_OUT_3
0            4369.0            5167.0            5599.0
1            4274.0            5108.0            5524.0
2            4301.0            5100.0            5532.0
3            4297.0            5111.0            5539.0
4            4296.0            5113.0            5535.0

In [12]:
# Boom Minus Y Wind Direction OUT = Number of counts measured for Wind Speed channel #
BMY_WD_OUT = df[['BMY_WD_OUT_1', 'BMY_WD_OUT_2', 'BMY_WD_OUT_3', 'BMY_WD_OUT_4', 'BMY_WD_OUT_5', 'BMY_WD_OUT_6', 'BMY_WD_OUT_7', 'BMY_WD_OUT_8', 'BMY_WD_OUT_9', 'BMY_WD_OUT_10', 'BMY_WD_OUT_11',  'BMY_WD_OUT_12']]
BMY_WD_OUT.head()

BMY_WD_OUT_1  BMY_WD_OUT_2  BMY_WD_OUT_3  BMY_WD_OUT_4  BMY_WD_OUT_5  \
0         648.0         662.0         648.0        1687.0        3360.0   
1         867.0         379.0         828.0        1033.0          10.0   
2         713.0         521.0         641.0        1085.0        1194.0   
3         687.0         578.0         662.0        1028.0        1254.0   
4         438.0         284.0         599.0         862.0        1085.0   

   BMY_WD_OUT_6  BMY_WD_OUT_7  BMY_WD_OUT_8  BMY_WD_OUT_9  BMY_WD_OUT_10  \
0        1109.0        1019.0        1229.0         868.0         2451.0   
1         551.0         578.0         324.0         637.0          418.0   
2         855.0         807.0         918.0         859.0         1233.0   
3         861.0         857.0         991.0         879.0         1190.0   
4         748.0         969.0        1049.0         813.0         1097.0   

   BMY_WD_OUT_11  BMY_WD_OUT_12  
0          760.0          663.0  
1          526.0          492.0  
2          741.0          661.0  
3          717.0          689.0  
4          417.0          443.0

In [13]:
frequency = df[['BMY_WIND_FREQUENCY', 'BMY_AIR_TEMP_FREQUENCY']]
BMY_ASIC_TEMP = df.BMY_ASIC_TEMP

In [14]:
BPY_2L_TEMP = df[['BPY_2L_TEMP_1', 'BPY_2L_TEMP_2', 'BPY_2L_TEMP_3', 'BPY_2L_TEMP_4', 'BPY_2L_TEMP_5', 'BPY_2L_TEMP_6']]
BPY_AIR_TEMP = df.BPY_AIR_TEMP
BPY_WD_REF_OUT = df[['BPY_WD_REF_OUT_1', 'BPY_WD_REF_OUT_2', 'BPY_WD_REF_OUT_3']]
BPY_WD_OUT = df[['BPY_WD_OUT_1', 'BPY_WD_OUT_2', 'BPY_WD_OUT_3', 'BPY_WD_OUT_4', 'BPY_WD_OUT_5', 'BPY_WD_OUT_6', 'BPY_WD_OUT_7', 'BPY_WD_OUT_8', 'BPY_WD_OUT_9', 'BPY_WD_OUT_10', 'BPY_WD_OUT_11', 'BPY_WD_OUT_12']]
BPY_FREQUENCY = df[['BPY_WIND_FREQUENCY', 'BPY_AIR_TEMP_FREQUENCY']]
BPY_ASIC_TEMP = df.BPY_ASIC_TEMP


In [15]:
df.BMY_AIR_TEMP_FREQUENCY.value_counts()

0.1    732339
0.5    690182
1.0     17480
Name: BMY_AIR_TEMP_FREQUENCY, dtype: int64

In [16]:
# I don't know why temp 6 is a positive number
# Temp 6 is for the Calibration resistor: 1K ohm
# I will drop it from the average 
BMY_2L_Temp['BMY_2L_TEMP_AVG'] = (BMY_2L_Temp.BMY_2L_TEMP_1 + BMY_2L_Temp.BMY_2L_TEMP_2 + BMY_2L_Temp.BMY_2L_TEMP_3 + BMY_2L_Temp.BMY_2L_TEMP_4 + BMY_2L_Temp.BMY_2L_TEMP_5)/5
BMY_2L_Temp.head()


BMY_2L_TEMP_1  BMY_2L_TEMP_2  BMY_2L_TEMP_3  BMY_2L_TEMP_4  BMY_2L_TEMP_5  \
0        -4353.0        -4645.0        -4778.0        -5006.0        -4512.0   
1        -4415.0        -4723.0        -4870.0        -5099.0        -4600.0   
2        -4424.0        -4727.0        -4861.0        -5107.0        -4613.0   
3        -4409.0        -4722.0        -4859.0        -5109.0        -4603.0   
4        -4424.0        -4728.0        -4873.0        -5107.0        -4610.0   

   BMY_2L_TEMP_6  BMY_2L_TEMP_AVG  
0         1002.0          -4658.8  
1          936.0          -4741.4  
2          930.0          -4746.4  
3          916.0          -4740.4  
4          922.0          -4748.4

In [17]:
# Wind Speed transducer cold die temperature
# I am going to make a column for the average of the 3 transducers
BMY_WD_REF_OUT['BMY_WD_REF_OUT_AVG'] = (BMY_WD_REF_OUT.BMY_WD_REF_OUT_1 + BMY_WD_REF_OUT.BMY_WD_REF_OUT_2 + BMY_WD_REF_OUT.BMY_WD_REF_OUT_3)/3
BMY_WD_REF_OUT.head()


BMY_WD_REF_OUT_1  BMY_WD_REF_OUT_2  BMY_WD_REF_OUT_3  BMY_WD_REF_OUT_AVG
0            4369.0            5167.0            5599.0         5045.000000
1            4274.0            5108.0            5524.0         4968.666667
2            4301.0            5100.0            5532.0         4977.666667
3            4297.0            5111.0            5539.0         4982.333333
4            4296.0            5113.0            5535.0         4981.333333

In [18]:
# should i keep them separate or combine them with a loop where it takes the column without null 
# or averages if both columns have data?
temp = pd.concat([BMY_AIR_TEMP , BPY_AIR_TEMP], axis=1)
temp.head()

BMY_AIR_TEMP  BPY_AIR_TEMP
0       -5703.0           NaN
1       -5776.0           NaN
2       -5803.0           NaN
3       -5771.0           NaN
4       -5754.0           NaN

In [19]:
# This is wind direction. I'm not sure I can calculate this in my MVP, so I think I'll leave this out for now
# I'll come back for this later
BMY_WD_OUT.head()

BMY_WD_OUT_1  BMY_WD_OUT_2  BMY_WD_OUT_3  BMY_WD_OUT_4  BMY_WD_OUT_5  \
0         648.0         662.0         648.0        1687.0        3360.0   
1         867.0         379.0         828.0        1033.0          10.0   
2         713.0         521.0         641.0        1085.0        1194.0   
3         687.0         578.0         662.0        1028.0        1254.0   
4         438.0         284.0         599.0         862.0        1085.0   

   BMY_WD_OUT_6  BMY_WD_OUT_7  BMY_WD_OUT_8  BMY_WD_OUT_9  BMY_WD_OUT_10  \
0        1109.0        1019.0        1229.0         868.0         2451.0   
1         551.0         578.0         324.0         637.0          418.0   
2         855.0         807.0         918.0         859.0         1233.0   
3         861.0         857.0         991.0         879.0         1190.0   
4         748.0         969.0        1049.0         813.0         1097.0   

   BMY_WD_OUT_11  BMY_WD_OUT_12  
0          760.0          663.0  
1          526.0          492.0  
2          741.0          661.0  
3          717.0          689.0  
4          417.0          443.0

In [20]:
# Air temperature channels frequency or frequencies
# Wind channels frequency or frequencies
# I don't think I need to know what frequency channel they were using at the time
# Not using in MVP 
frequency.head()

BMY_WIND_FREQUENCY  BMY_AIR_TEMP_FREQUENCY
0                 1.0                     1.0
1                 1.0                     1.0
2                 1.0                     1.0
3                 1.0                     1.0
4                 1.0                     1.0

In [21]:
BMY_simple_df = pd.concat([BMY_2L_Temp.BMY_2L_TEMP_AVG, # temperature
                           BMY_AIR_TEMP, # temperature
                           BMY_WD_REF_OUT.BMY_WD_REF_OUT_AVG, # temperature
                           BMY_ASIC_TEMP # temperature
                          ], axis=1)

BMY_simple_df.head()


BMY_2L_TEMP_AVG  BMY_AIR_TEMP  BMY_WD_REF_OUT_AVG  BMY_ASIC_TEMP
0          -4658.8       -5703.0         5045.000000         8485.0
1          -4741.4       -5776.0         4968.666667         8488.0
2          -4746.4       -5803.0         4977.666667         8491.0
3          -4740.4       -5771.0         4982.333333         8494.0
4          -4748.4       -5754.0         4981.333333         8498.0

In [22]:
# BPY_2L_TEMP_4 is the resistor, so I'll skip that for averaging again
BPY_2L_TEMP['BPY_2L_TEMP_AVG'] = (BPY_2L_TEMP.BPY_2L_TEMP_1 + BPY_2L_TEMP.BPY_2L_TEMP_2 + BPY_2L_TEMP.BPY_2L_TEMP_3 + BPY_2L_TEMP.BPY_2L_TEMP_5 + BPY_2L_TEMP.BPY_2L_TEMP_6)/5
BPY_2L_TEMP.head(100)


BPY_2L_TEMP_1  BPY_2L_TEMP_2  BPY_2L_TEMP_3  BPY_2L_TEMP_4  BPY_2L_TEMP_5  \
0             NaN            NaN            NaN            NaN            NaN   
1             NaN            NaN            NaN            NaN            NaN   
2             NaN            NaN            NaN            NaN            NaN   
3             NaN            NaN            NaN            NaN            NaN   
4             NaN            NaN            NaN            NaN            NaN   
..            ...            ...            ...            ...            ...   
95        -2915.0        -2463.0        -2062.0         1682.0        -2242.0   
96        -2896.0        -2453.0        -2055.0         1703.0        -2228.0   
97        -2887.0        -2441.0        -2054.0         1709.0        -2229.0   
98        -2908.0        -2444.0        -2048.0         1685.0        -2218.0   
99        -2872.0        -2422.0        -2039.0         1699.0        -2210.0   

    BPY_2L_TEMP_6  BPY_2L_TEMP_AVG  
0             NaN              NaN  
1             NaN              NaN  
2             NaN              NaN  
3             NaN              NaN  
4             NaN              NaN  
..            ...              ...  
95        -2201.0          -2376.6  
96        -2193.0          -2365.0  
97        -2177.0          -2357.6  
98        -2185.0          -2360.6  
99        -2167.0          -2342.0  

[100 rows x 7 columns]

In [23]:
# 
BPY_WD_REF_OUT['BPY_WD_REF_OUT_AVG'] = (BPY_WD_REF_OUT.BPY_WD_REF_OUT_1 + BPY_WD_REF_OUT.BPY_WD_REF_OUT_2 + BPY_WD_REF_OUT.BPY_WD_REF_OUT_3)/3

BPY_WD_REF_OUT.head(100)


BPY_WD_REF_OUT_1  BPY_WD_REF_OUT_2  BPY_WD_REF_OUT_3  BPY_WD_REF_OUT_AVG
0                NaN               NaN               NaN                 NaN
1                NaN               NaN               NaN                 NaN
2                NaN               NaN               NaN                 NaN
3                NaN               NaN               NaN                 NaN
4                NaN               NaN               NaN                 NaN
..               ...               ...               ...                 ...
95           10521.0           10048.0           10397.0        10322.000000
96           10521.0           10048.0           10401.0        10323.333333
97           10522.0           10042.0           10409.0        10324.333333
98           10532.0           10019.0           10407.0        10319.333333
99           10562.0           10037.0           10405.0        10334.666667

[100 rows x 4 columns]

In [24]:
BPY_simple_df = pd.concat([BPY_2L_TEMP.BPY_2L_TEMP_AVG, # temperature
                           BPY_AIR_TEMP, # temperature
                           BPY_WD_REF_OUT.BPY_WD_REF_OUT_AVG, # temperature
                           BPY_ASIC_TEMP # temperature
                          ], axis=1)
BPY_simple_df.head(100)

BPY_2L_TEMP_AVG  BPY_AIR_TEMP  BPY_WD_REF_OUT_AVG  BPY_ASIC_TEMP
0               NaN           NaN                 NaN            NaN
1               NaN           NaN                 NaN            NaN
2               NaN           NaN                 NaN            NaN
3               NaN           NaN                 NaN            NaN
4               NaN           NaN                 NaN            NaN
..              ...           ...                 ...            ...
95          -2376.6       -2238.0        10322.000000         8886.0
96          -2365.0       -2216.0        10323.333333         8892.0
97          -2357.6       -2202.0        10324.333333         8895.0
98          -2360.6       -2227.0        10319.333333         8900.0
99          -2342.0       -2221.0        10334.666667         8904.0

[100 rows x 4 columns]

In [25]:
df.head()

AOBT          SCLK                LMST            LTST  \
0  596876952.0  5.968612e+08  00004M06:46:33.826  00004 06:05:41   
1  596876953.0  5.968612e+08  00004M06:46:34.799  00004 06:05:42   
2  596876954.0  5.968612e+08  00004M06:46:35.772  00004 06:05:43   
3  596876955.0  5.968612e+08  00004M06:46:36.746  00004 06:05:44   
4  596876956.0  5.968612e+08  00004M06:46:37.719  00004 06:05:45   

                      UTC  BMY_2L_TEMP_1  BMY_2L_TEMP_2  BMY_2L_TEMP_3  \
0  2018-334T14:46:55.755Z        -4353.0        -4645.0        -4778.0   
1  2018-334T14:46:56.755Z        -4415.0        -4723.0        -4870.0   
2  2018-334T14:46:57.755Z        -4424.0        -4727.0        -4861.0   
3  2018-334T14:46:58.755Z        -4409.0        -4722.0        -4859.0   
4  2018-334T14:46:59.755Z        -4424.0        -4728.0        -4873.0   

   BMY_2L_TEMP_4  BMY_2L_TEMP_4_AVERAGE  ...  BPY_WD_OUT_6  BPY_WD_OUT_7  \
0        -5006.0                    NaN  ...           NaN           NaN   
1        -5099.0                    NaN  ...           NaN           NaN   
2        -5107.0                    NaN  ...           NaN           NaN   
3        -5109.0                    NaN  ...           NaN           NaN   
4        -5107.0                    NaN  ...           NaN           NaN   

   BPY_WD_OUT_8  BPY_WD_OUT_9  BPY_WD_OUT_10  BPY_WD_OUT_11  BPY_WD_OUT_12  \
0           NaN           NaN            NaN            NaN            NaN   
1           NaN           NaN            NaN            NaN            NaN   
2           NaN           NaN            NaN            NaN            NaN   
3           NaN           NaN            NaN            NaN            NaN   
4           NaN           NaN            NaN            NaN            NaN   

   BPY_WIND_FREQUENCY  BPY_AIR_TEMP_FREQUENCY  BPY_ASIC_TEMP  
0                 NaN                     NaN            NaN  
1                 NaN                     NaN            NaN  
2                 NaN                     NaN            NaN  
3                 NaN                     NaN            NaN  
4                 NaN                     NaN            NaN  

[5 rows x 63 columns]

# Combining BMY and BPY

<hr style="border:2px solid red"> </hr>

In [26]:
# Is there an overlap where BMY and BPY are collecting data at the same time?
# YES!
df[['BMY_2L_TEMP_1', 'BPY_2L_TEMP_1']].isnull().sum(axis=1).value_counts()

1    1821589
0     528809
2         38
dtype: int64

Translation:  
True + False = 1: There is only one column that has data (is not null)  
False + False = 0: Both columns have data, (niether is null)  
True + True = 2: Neither column has any recorded data (both are null)

In [27]:
# I want to have a continuous column of data.
# Since BMP and BPY represent InSight alternating sides to collect data, I need to make a function
# When they are both recording data I will take the average
# Otherwise I will record whatever side has data
# if no data is present I will keep the null value so I can see the times where there are gaps

temp_df = df[['BMY_2L_TEMP_1', 'BPY_2L_TEMP_1']]
BMY_array = temp_df.BMY_2L_TEMP_1.to_numpy()
BPY_array = temp_df.BPY_2L_TEMP_1.to_numpy()
#temp_arrays = np.stack([BMY_array,BPY_array])
mars_temp = []

# for x, y in zip(array1, array2):   
for BMY, BPY in zip(BMY_array, BPY_array):
    
    # if neither value is null, take the mean
    if np.isnan(BMY) != True & np.isnan(BPY) != True :
        mars_temp.append((BMY + BPY)/2)

    # if BMY is only column with data, add to list
    elif np.isnan(BMY) != True & np.isnan(BPY) == True :
        mars_temp.append(BMY)

    # if BPY is only column with data, add to list
    elif np.isnan(BMY) == True & np.isnan(BPY) != True :
        mars_temp.append(BPY)
    
    else:
        mars_temp.append(BMY)

mars_temp = pd.DataFrame(mars_temp)
mars_temp.head()

0
0 -4353.0
1 -4415.0
2 -4424.0
3 -4409.0
4 -4424.0

In [68]:
# I need to make this a function!
# Output is a dataframe
def combine_TWINS(BMY_array, BPY_array):
    # Create the list to catch our data
    mars_temp = []

    # for x, y in zip(array1, array2):   
    for BMY, BPY in zip(BMY_array, BPY_array):

        # If both columns have a value, take the average of the two
        if (np.isnan(BMY) + np.isnan(BPY)) == 0 :
            mars_temp.append((BMY + BPY)/2)
        
        # If both columns have null value, add the first null value
        elif (np.isnan(BMY) + np.isnan(BPY)) == 2:
            mars_temp.append(BMY)

        # If BMY is only column with data, add to list
        elif np.isnan(BMY) == False:
            mars_temp.append(BMY)
        
        # If BPY is only column with data, add to list
        else:
            mars_temp.append(BPY)

    return pd.DataFrame(mars_temp)


In [64]:
np.isnan(exampleA[0]) + False == 0

True

In [69]:
# Debugging
mars_temp = []
mars_temp.append((exampleA[0] + exampleB[0])/2)
mars_temp.append(exampleA[1])

mars_temp = pd.DataFrame(mars_temp)

mars_temp

0
0 -10461.3
1 -10712.0

## DEBUG
<hr style="border:2px solid red"> </hr>

I have more nulls in my Final Dataframe below, than I started with!

In [29]:
# From the dataframes I made that averaged the data: How much data is missing in 2L_TEMP_AVG?
pd.concat([BMY_simple_df.BMY_2L_TEMP_AVG, BPY_simple_df.BPY_2L_TEMP_AVG], axis=1).isnull().sum(axis=1).value_counts()


2    1251752
1     782775
0     315909
dtype: int64

Translation:  
True + False = 1: There is only one column that has data (is not null)  
False + False = 0: Both columns have data, (niether is null)  
True + True = 2: Neither column has any recorded data (both are null)

Concerning 2L_TEMP_AVG: There are 1,098,684 rows with data in them, and 1,251,752 with only nulls. Even though there are 2,350,436 rows, only <b><u>46.74%</b></u> of these rows have data.

In [30]:
# checking based on original df
pd.concat([df.BMY_AIR_TEMP, df.BPY_AIR_TEMP], axis=1).isnull().sum(axis=1).value_counts()


2    1251752
1     782775
0     315909
dtype: int64

In [31]:
# Do these columns have data in the same place?
# They do! Wohoo!
pd.concat([BPY_simple_df.BPY_2L_TEMP_AVG, BPY_simple_df.BPY_AIR_TEMP], axis=1).isnull().sum(axis=1).value_counts()


2    1643505
0     706931
dtype: int64

Concerning AIR_TEMP: There are 1,098,684 rows with data in them, and 1,251,752 with only nulls. Even though there are 2,350,436 rows, only <b><u>46.74%</b></u> of these rows have data.

In [32]:
# From the dataframes I made that averaged the data: How much data is missing in ASIC_TEMP?
pd.concat([BMY_simple_df.BMY_ASIC_TEMP, BPY_simple_df.BPY_ASIC_TEMP], axis=1).isnull().sum(axis=1).value_counts()


1    1821613
0     528806
2         17
dtype: int64

In [33]:
# Checking myself again
pd.concat([df.BMY_ASIC_TEMP, df.BPY_ASIC_TEMP], axis=1).isnull().sum(axis=1).value_counts()


1    1821613
0     528806
2         17
dtype: int64

Concerning ASIC_TEMP: There are 2,350,419 rows with data in them, and 17 with only nulls. There are 2,350,436 rows, and <b><u>99.99%</b></u> of these rows have data.  
Why does this have more data? What is ASIC?  
ASIC stands for Application-Specific Integrated Circuit. I assume this means the temperature of that circuit board? So I don't want to use this!

In [70]:
exampleA = BPY_simple_df.BPY_2L_TEMP_AVG[140:160].to_numpy()
exampleB = BMY_simple_df.BMY_2L_TEMP_AVG[140:160].to_numpy()
exampleC= combine_TWINS(exampleA, exampleB)
print(exampleA)
print(exampleB)
print(exampleC)

[-10737.4 -10712.  -10702.8 -10687.8 -10713.8 -10709.2 -10749.6 -10627.
 -10774.       nan      nan      nan      nan      nan      nan  -2175.8
  -2280.2  -2273.8  -2273.4  -2256.4]
[-10185.2 -10251.2 -10182.8      nan      nan      nan      nan      nan
      nan  -4478.6  -4563.   -4574.8  -4569.4  -4574.6  -4555.6  -4551.8
  -4562.8  -4556.8  -4543.2  -4548.6]
          0
0  -10461.3
1  -10481.6
2  -10442.8
3  -10687.8
4  -10713.8
5  -10709.2
6  -10749.6
7  -10627.0
8  -10774.0
9   -4478.6
10  -4563.0
11  -4574.8
12  -4569.4
13  -4574.6
14  -4555.6
15  -3363.8
16  -3421.5
17  -3415.3
18  -3408.3
19  -3402.5


<hr style="border:2px solid red"> </hr>

In [71]:
# I want my dataframe to look like this: TIME | Temp | Air Temp | ASIC Temp
# I will use the dataframes I made that averaged all the columns
# I will start by taking a designated column (e.g."air_temp") 
# and send it to my function as two arrays, one from BMY and one from BPY
# I should then get one column representing both dataframes

# 2L_TEMP
twins_2L_TEMP = combine_TWINS(BMY_simple_df.BMY_2L_TEMP_AVG.to_numpy(), 
                              BPY_simple_df.BPY_2L_TEMP_AVG.to_numpy() )

# Air_TEMP
twins_air_TEMP = combine_TWINS(BMY_simple_df.BMY_AIR_TEMP.to_numpy(), 
                               BPY_simple_df.BPY_AIR_TEMP.to_numpy() )



In [73]:
# Let's see if that helped our null values
twins_2L_TEMP.isnull().sum(axis=1).value_counts()



1    1251752
0    1098684
dtype: int64

Concerning 2L_TEMP_AVG: There are 1,098,684 rows with data in them, and 1,251,752 with only nulls. Even though there are 2,350,436 rows, only 46.74% of these rows have data.

In [74]:
# Let's see if that helped our null values
twins_air_TEMP.isnull().sum(axis=1).value_counts()


1    1251752
0    1098684
dtype: int64

In [72]:
df.head()

AOBT          SCLK                LMST            LTST  \
0  596876952.0  5.968612e+08  00004M06:46:33.826  00004 06:05:41   
1  596876953.0  5.968612e+08  00004M06:46:34.799  00004 06:05:42   
2  596876954.0  5.968612e+08  00004M06:46:35.772  00004 06:05:43   
3  596876955.0  5.968612e+08  00004M06:46:36.746  00004 06:05:44   
4  596876956.0  5.968612e+08  00004M06:46:37.719  00004 06:05:45   

                      UTC  BMY_2L_TEMP_1  BMY_2L_TEMP_2  BMY_2L_TEMP_3  \
0  2018-334T14:46:55.755Z        -4353.0        -4645.0        -4778.0   
1  2018-334T14:46:56.755Z        -4415.0        -4723.0        -4870.0   
2  2018-334T14:46:57.755Z        -4424.0        -4727.0        -4861.0   
3  2018-334T14:46:58.755Z        -4409.0        -4722.0        -4859.0   
4  2018-334T14:46:59.755Z        -4424.0        -4728.0        -4873.0   

   BMY_2L_TEMP_4  BMY_2L_TEMP_4_AVERAGE  ...  BPY_WD_OUT_6  BPY_WD_OUT_7  \
0        -5006.0                    NaN  ...           NaN           NaN   
1        -5099.0                    NaN  ...           NaN           NaN   
2        -5107.0                    NaN  ...           NaN           NaN   
3        -5109.0                    NaN  ...           NaN           NaN   
4        -5107.0                    NaN  ...           NaN           NaN   

   BPY_WD_OUT_8  BPY_WD_OUT_9  BPY_WD_OUT_10  BPY_WD_OUT_11  BPY_WD_OUT_12  \
0           NaN           NaN            NaN            NaN            NaN   
1           NaN           NaN            NaN            NaN            NaN   
2           NaN           NaN            NaN            NaN            NaN   
3           NaN           NaN            NaN            NaN            NaN   
4           NaN           NaN            NaN            NaN            NaN   

   BPY_WIND_FREQUENCY  BPY_AIR_TEMP_FREQUENCY  BPY_ASIC_TEMP  
0                 NaN                     NaN            NaN  
1                 NaN                     NaN            NaN  
2                 NaN                     NaN            NaN  
3                 NaN                     NaN            NaN  
4                 NaN                     NaN            NaN  

[5 rows x 63 columns]

In [78]:
local_true_solar_time = pd.DataFrame(df.LTST)
local_true_solar_time.reset_index(inplace=True, drop=True)
local_true_solar_time.head()

LTST
0  00004 06:05:41
1  00004 06:05:42
2  00004 06:05:43
3  00004 06:05:44
4  00004 06:05:45

In [75]:
# Checking my dataframe before merge
twins_2L_TEMP.rename(columns={0:'twins_2L_TEMP'}, inplace=True)
twins_2L_TEMP.head()

twins_2L_TEMP
0        -4658.8
1        -4741.4
2        -4746.4
3        -4740.4
4        -4748.4

In [76]:
# Checking my dataframe before merge
twins_air_TEMP.rename(columns={0:'twins_air_TEMP'}, inplace=True)
twins_air_TEMP.head()

twins_air_TEMP
0         -5703.0
1         -5776.0
2         -5803.0
3         -5771.0
4         -5754.0

In [37]:
# Checking my dataframe before merge
twins_ASIC_TEMP.rename(columns={0:'twins_ASIC_TEMP'}, inplace=True)
twins_ASIC_TEMP.head()

twins_ASIC_TEMP
0           8485.0
1           8488.0
2           8491.0
3           8494.0
4           8498.0

In [51]:
error = pd.concat([local_true_solar_time, twins_2L_TEMP], axis=1, ignore_index=True, sort=False)
error.head()

0       1
0  00004 06:05:41 -4658.8
1  00004 06:05:42 -4741.4
2  00004 06:05:43 -4746.4
3  00004 06:05:44 -4740.4
4  00004 06:05:45 -4748.4

In [79]:
final_df = pd.concat([local_true_solar_time,twins_2L_TEMP, twins_air_TEMP], axis=1, ignore_index=True, sort=False)
final_df.rename(columns={0:'local_true_solar_time', 1:'twins_2L_TEMP', 2:'twins_air_TEMP'}, inplace=True)

final_df.head()


local_true_solar_time  twins_2L_TEMP  twins_air_TEMP
0        00004 06:05:41        -4658.8         -5703.0
1        00004 06:05:42        -4741.4         -5776.0
2        00004 06:05:43        -4746.4         -5803.0
3        00004 06:05:44        -4740.4         -5771.0
4        00004 06:05:45        -4748.4         -5754.0

<hr style="border:2px solid red"> </hr>

In [80]:
# Now I have to make sure my final dataframe has a datetime
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2350436 entries, 0 to 2350435
Data columns (total 3 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   local_true_solar_time  object 
 1   twins_2L_TEMP          float64
 2   twins_air_TEMP         float64
dtypes: float64(2), object(1)
memory usage: 53.8+ MB


In [81]:
#final_df.local_true_solar_time = final_df.local_true_solar_time.str[2:]
final_df.local_true_solar_time

0          00004 06:05:41
1          00004 06:05:42
2          00004 06:05:43
3          00004 06:05:44
4          00004 06:05:45
                ...      
2350431    00120 05:19:07
2350432    00120 05:19:12
2350433    00120 05:19:17
2350434    00120 05:19:21
2350435    00120 05:19:26
Name: local_true_solar_time, Length: 2350436, dtype: object

In [75]:
# This is going to be tricky, because datetime uses earth values. I need to make sure it keeps Mars time!
# final_df.local_true_solar_time = pd.to_datetime(final_df.local_true_solar_time, format = '%j %H:%M:%S')
# final_df.local_true_solar_time.head()

0   1900-01-04 06:05:41
1   1900-01-04 06:05:42
2   1900-01-04 06:05:43
3   1900-01-04 06:05:44
4   1900-01-04 06:05:45
Name: local_true_solar_time, dtype: datetime64[ns]

In [82]:
def mars_to_earth_time(mars_date):
    mars_day = int(mars_date[:5])
    mars_hour = int(mars_date[6:8])
    mars_minute = int(mars_date[9:11])
    mars_second = int(mars_date[12:14])

    # mars_days relates to how many sol days InSight has been on Mars
    # InSight landed on Mars on November 26, 2018 2:52:59 PM EST (Earth Time)
    earth_time = landing_day = pd.to_datetime('November 26, 2018 14:52:59', format = '%B %d, %Y %H:%M:%S')

    # 1 mars day = 24 hours, 39 minutes and 35 seconds on Earth
    # For every mars day, add that total time to earth_time
    for x in range(int(mars_day)):
        earth_time += datetime.timedelta(hours= 24, minutes= 39, seconds= 35)
    
    # Finally, add the remaining hours, minutes, and seconds
    earth_time += datetime.timedelta(hours= mars_hour, minutes= mars_minute, seconds= mars_second)
    
    return earth_time


In [83]:
# Testing my new function before converting my dataframe
print('Mars Date:', final_df.local_true_solar_time[0])

print('Earth Date:', mars_to_earth_time(final_df.local_true_solar_time[0]))


Mars Date: 00004 06:05:41
Earth Date: 2018-11-30 23:37:00


In [85]:
# This cell takes about 16 minutes to run
# for every local_true_solar_time, convert to earth_date
final_df['earth_date']= final_df.local_true_solar_time.apply(mars_to_earth_time)
final_df.head()

local_true_solar_time  twins_2L_TEMP  twins_air_TEMP          earth_date
0        00004 06:05:41        -4658.8         -5703.0 2018-11-30 23:37:00
1        00004 06:05:42        -4741.4         -5776.0 2018-11-30 23:37:01
2        00004 06:05:43        -4746.4         -5803.0 2018-11-30 23:37:02
3        00004 06:05:44        -4740.4         -5771.0 2018-11-30 23:37:03
4        00004 06:05:45        -4748.4         -5754.0 2018-11-30 23:37:04

In [86]:
# Since converting the time took so long, I'm gonna make sure I have a quick and easy back-up!
final_df.to_csv('MVP2.csv')

In [87]:
# Now to make a function that will give me exactly what I want to use in the Explore stage
df = pd.read_csv('MVP2.csv').reset_index(drop=True)
df = df.drop(columns=['Unnamed: 0', 'local_true_solar_time'])
df.head()

twins_2L_TEMP  twins_air_TEMP           earth_date
0        -4658.8         -5703.0  2018-11-30 23:37:00
1        -4741.4         -5776.0  2018-11-30 23:37:01
2        -4746.4         -5803.0  2018-11-30 23:37:02
3        -4740.4         -5771.0  2018-11-30 23:37:03
4        -4748.4         -5754.0  2018-11-30 23:37:04

In [88]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2350436 entries, 0 to 2350435
Data columns (total 3 columns):
 #   Column          Dtype  
---  ------          -----  
 0   twins_2L_TEMP   float64
 1   twins_air_TEMP  float64
 2   earth_date      object 
dtypes: float64(2), object(1)
memory usage: 53.8+ MB


In [89]:
# Converting earth_date to datetime dtype
df.earth_date = pd.to_datetime(df.earth_date, format = '%Y-%m-%d %H:%M:%S')
df.earth_date.head()

0   2018-11-30 23:37:00
1   2018-11-30 23:37:01
2   2018-11-30 23:37:02
3   2018-11-30 23:37:03
4   2018-11-30 23:37:04
Name: earth_date, dtype: datetime64[ns]

In [90]:
# Setting the 'earth_date' column as the Index and sorting that new Index:
df = df.set_index('earth_date').sort_index()
df.head()

twins_2L_TEMP  twins_air_TEMP
earth_date                                        
2018-11-30 23:37:00        -4658.8         -5703.0
2018-11-30 23:37:01        -4741.4         -5776.0
2018-11-30 23:37:02        -4746.4         -5803.0
2018-11-30 23:37:03        -4740.4         -5771.0
2018-11-30 23:37:04        -4748.4         -5754.0

In [91]:
# double checking my null values
df.isnull().sum()

twins_2L_TEMP     1251752
twins_air_TEMP    1251752
dtype: int64

In [92]:
df.shape

(2350436, 2)

In [94]:
# Percentages of missing data
print('twins_2L_TEMP:', df.twins_2L_TEMP.isnull().sum()/df.shape[0])
print('twins_air_TEMP', df.twins_air_TEMP.isnull().sum()/df.shape[0])


twins_2L_TEMP: 0.5325616183550627
twins_air_TEMP 0.5325616183550627


In [96]:
df.head()

twins_2L_TEMP  twins_air_TEMP
earth_date                                        
2018-11-30 23:37:00        -4658.8         -5703.0
2018-11-30 23:37:01        -4741.4         -5776.0
2018-11-30 23:37:02        -4746.4         -5803.0
2018-11-30 23:37:03        -4740.4         -5771.0
2018-11-30 23:37:04        -4748.4         -5754.0

In [95]:
def prepare_TWINS():
    # Read in data and drop unneeded columns
    df = pd.read_csv('MVP2.csv').reset_index(drop=True)
    df = df.drop(columns=['Unnamed: 0', 'local_true_solar_time'])

    # Converting earth_date to datetime dtype
    df.earth_date = pd.to_datetime(df.earth_date, format = '%Y-%m-%d %H:%M:%S')

    # Setting the 'earth_date' column as the Index and sorting that new Index:
    df = df.set_index('earth_date').sort_index()
    
    return df